docs/data-sourcing-and-transformation.md

```diff
@@ -0,0 +1,91 @@
+# Data Sourcing and Transformations
+
+![alpha](assets/alpha.svg)
+
+> v3.0 and after
+
+#### Development
+
+We have intentionally made this feature available with only bare-bones functionality. Our hope is that we are able to build this feature with our community's feedback. If you have ideas and use cases for this feature, please open an [enhancement proposal](https://github.com/argoproj/argo-workflows/issues/new?assignees=&labels=enhancement&template=enhancement_proposal.md) on GitHub.
+
+Aditionally, please take a look at our current ideas at the bottom of this document.
+
+## Introduction
+Users often source and transform data as part of their workflows. The `data` template provides first-class support for these common operations.
+
+`data` templates can best be understood by looking at a common data sourcing and transformation operation in `bash`:
+
+```bash
+$ find -r . | grep ".pdf" | sed "s/foo/foo.ready/"
+```
+
+Such operations consist of two main parts:
+* A "source" of data: `find -r .`
+* A series of "transformations" which transform the output of the source serially: `| grep ".pdf" | sed "s/foo/foo.ready/"`
+
+This operation, for example, could be useful in sourcing a potential list of files to be processed and filtering and manipulating the list as desired.
+
+In Argo, this operation would be written as:
+
+```yaml
+- name: generate-artifacts
+  data:
+    source:             # Define a source for the data, only a single "source" is permitted
+      artifactPaths:    # A predefined source: Generate a list of all artifact paths in a given repository
+        s3:             # Source from an S3 bucket
+          bucket: test
+          endpoint: minio:9000
+          insecure: true
+          accessKeySecret:
+            name: my-minio-cred
+            key: accesskey
+          secretKeySecret:
+            name: my-minio-cred
+            key: secretkey
+    transformation:     # The source is then passed to be transformed by transformations defined here
+      - expression: "filter(data, {# endsWith \".pdf\"})"
+      - expression: "map(data, {# + \".ready\"})"
+```
+
+## In-Memory Transformations
+
+If possible, Argo will choose to perform the specified transformation in its own memory, saving the need to create a pod in the cluster for said transformation. Because of this, in-memory transformations are near instantaneously completed (much like a [cached/memoized node](memoization.md)).
+
+Currently, Argo will always choose to perform the transformation in memory if it does not use `artifactPaths` to source the data.
```

Comment:
this will change over time, so I would remove this line

Reply:
I intend to update this as time goes on. I think it's important that it's transparent to the user when they can expect Pods to not be created